### PR TITLE
Add Definition Presenter

### PR DESCRIPTION
**Pull request by ghickman**
This adds the Definition presentation function.  It converts a given Definition instance into a structure (list of dictionaries) ready for display in a template.  The template is now in charge of styling rows of a Definition.

This is a rebuild of #110 on top of @inglesp's new `Hierarchy` object.

### CODE REVIEW

File: codelists/presenters.py [hunk 1]
```diff
@@ -1,8 +1,72 @@
 from itertools import groupby
 
+import attr
 from django.urls import reverse
 
 
+@attr.s
+class Row:
```
Comment (inglesp) [inline codelists/presenters.py:8]: Could we call this `DefinitionRow`?

File: codelists/presenters.py [hunk 2]
```diff
@@ -1,8 +1,72 @@
...
+
+        * all descendants are included (all_descendants = True, excluded_children = [])
+        * all descendants except N (all_descendants = True, excluded_children = [...])
+        * no descendants (all_descendants = False, excluded_children ignored)
+
+    """
+
+    name: str = attr.ib()
+    code: str = attr.ib()
+    excluded_children: list = attr.ib(default=list())
```
Comment (inglesp) [inline codelists/presenters.py:28]: Can you change this to `excluded_descendants`?

File: codelists/views.py [hunk 3]
```diff
@@ -311,7 +311,7 @@ def version(request, project_slug, codelist_slug, qualified_version_str):
         "headers": headers,
         "rows": rows,
         "html_tree": html_tree,
-        "html_definition": html_definition,
+        "definition": definition_rows,
```
Comment (inglesp) [inline codelists/views.py:314]: Could the key be `"definition_rows"`?

File: codelists/presenters.py [hunk 4]
```diff
@@ -1,8 +1,72 @@
...
+    all_descendants: bool = attr.ib(default=True)
+
+
+def _iter_rules(hierarchy, rules, name_for_rule, excluded=None):
```
Comment (inglesp) [inline codelists/presenters.py:32]: This works while a `DefinitionRule` does not have both `applies_to_descendants=True` and `code_is_excluded=True`.  That is currently the case because of how `Definition.from_codes` works.

If this wasn't the case, we'd end up recursing a second time (ie calling the function three times) which would mean that `r in excluded` raises a `TypeError` because `excluded` is `None`.

I wonder whether it's worth adding a check (an `assert` would do) to `DefinitionRule.__init__` to ensure that the the invariant is maintained.

File: codelists/presenters.py [hunk 5]
```diff
@@ -1,8 +1,72 @@
...
+    """
+    Data structure for a Definition to prepare for display
+
+    name: name of the definition
+    code: code of the definition
+    excluded_children: list of excluded children (created with this class too)
+    all_descendants: are all of this Definitions descendants included?
+
+    Between all_descendants and excluded_children we can cover the three state
+    situation a Definition can be in:
```
Comment (inglesp) [inline codelists/presenters.py:18]: You're using `Definition` but I think you mean `DefinitionRule`.

File: codelists/templatetags/definitions.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+from django import template
+from django.urls import reverse
+
+register = template.Library()
+
+
+@register.simple_tag
+def concept_url(coding_system_id, code):
+    return reverse(f"{coding_system_id}:concept", args=[code])
```
Comment (inglesp) [inline codelists/templatetags/definitions.py:9]: I like this.

File: codelists/presenters.py [hunk 7]
```diff
@@ -1,8 +1,72 @@
...
+
+        # filter the excluded rules down to children of the current Rule
+        excluding_rules = [
+            r for r in excluded if r.code in hierarchy.descendants(rule.code)
+        ]
+
+        # generate excluded children by recursing into this function
+        row.excluded_children = list(
+            _iter_rules(hierarchy, excluding_rules, name_for_rule)
+        )
```
Comment (inglesp) [inline codelists/presenters.py:53]: I found the recursion a bit hard to follow -- in part because on the first call, `rules` are included rules, and on the second call (ie the first recursion) they are excluded rules.

Given that on the first recursion, we only yield `Row(name=name_for_rule(rule), code=rule.code, all_descendants=False)`, I'd be tempted to avoid recursing altogether, and build up `row.excluded_children` directly.

**Review by inglesp — APPROVED** [review]
Good stuff.  There are some small naming changes, and a thought about how you could avoid some of the recursion.

File: codelists/presenters.py (hunk 7 above)
Comment (ghickman) [reply codelists/presenters.py]: Great shout, this was very much a case of not seeing the wood for the trees.  Fixed in dad666a and much easier to read I think!

File: codelists/presenters.py (hunk 5 above)
Comment (ghickman) [reply codelists/presenters.py]: Fixed 8491893

File: codelists/presenters.py (hunk 1 above)
Comment (ghickman) [reply codelists/presenters.py]: Fixed in 8491893

File: codelists/presenters.py (hunk 2 above)
Comment (ghickman) [reply codelists/presenters.py]: Fixed in 8491893

File: codelists/views.py (hunk 3 above)
Comment (ghickman) [reply codelists/views.py]: Fixed in 8491893

File: codelists/presenters.py (hunk 4 above)
Comment (ghickman) [reply codelists/presenters.py]: Makes sense!  I've added it in c44bbcb as an exception so I can express why it happened and for what code (which should help traceability when we hit this $later).

Comment (inglesp) [reply codelists/presenters.py]: Great.
